src/shapes/IText/ITextBehavior.ts

```diff
@@ -616,6 +617,13 @@ export abstract class ITextBehavior<
     p.x += this.canvas._offset.left;
     p.y += this.canvas._offset.top;
 
+    const doc =
+      (this.canvas && getDocumentFromElement(this.canvas.getElement())) ||
+      getFabricDocument();
+    if (p.y > doc.body.clientHeight - charHeight) {
+      p.y = doc.body.clientHeight - charHeight;
```

Comment:
how this `doc.body.clientHeight - charHeight` relates to  `maxHeight = upperCanvasHeight - charHeight` that we check up before, and why we can't have a single max vertical limit and use that?

Reply:
https://github.com/fabricjs/fabric.js/pull/9948/files#diff-355c433500e45e82cda2cb4339613f1dacbb1d6ef43d8b0b98bf8fd293712976R588

Reply:
The `maxHeight` ensures that the position of the hiddenTextArea does not exceed the bounds of the canvas, which means the calculated absolute coordinates are relative to the canvas's position. However, our hiddenTextarea is appended to the document.body, and there's no problem as long as the canvas stays within the viewport. But if the canvas extends beyond the viewport, the absolutely positioned hiddenTextarea would exceed the viewport, causing the browser to shift all elements upward to ensure hiddenTextArea is visible on screen. So I add a final check to ensure the calcualted absolute coordinates not exceed the viewport.

Reply:
![image](https://github.com/fabricjs/fabric.js/assets/16117034/7654ef3a-be92-436d-9371-666d2e5d1b82)
The image on the left shows the scenario where maxHeight is effective, with hiddenTextarea displayed within the canvas boundary. The image on the right illustrates the situation where the canvas extends beyond the viewport; the hiddenTextarea meets the maxHeight condition, but it exceeds the range of the viewport, causing the browser to shift all elements upwards to ensure the hiddenTextArea is visible on the screen.

Reply:
Additionally, in my opinion, `maxHeight` should actually be expressed as `maxCanvasHeight`, and then add the canvas offset to get the true position:`p.y += this.canvas._offset.top;`. Therefore, I didn't use the same `maxHeight` to make the judgment, but instead added another judgment later to ensure that it doesn't affect the logic in between.

Reply:
Additionally, in my opinion, `maxHeight` should actually be expressed as `maxCanvasHeight`, and then add the canvas offset to get the true position:`p.y += this.canvas._offset.top;`. Therefore, I didn't use the same `maxHeight` to make the judgment, but instead added another judgment later to ensure that it doesn't affect the logic in between.

Reply:
This leaves me confused. If the hidden textarea is there, and i m editing a text that is below the fold, how i m going to see it?

Reply:
ToT. I'm sorry, my expression ability is poor.
Just use the link I shared,[https://jsfiddle.net/zacharyzch/L4sy3gu1/134/](https://jsfiddle.net/zacharyzch/L4sy3gu1/134/), if you continuously press Enter in the textarea, the red canvas will get pushed up because of the hiddenTextarea

Reply:
I understand the move page up issue. What i do not understand is if the textarea is below the fold and you don't move the page up, i tried your fiddle in a number of ways.
There are 2 issues here, one is that the page scrolls and one is that it scrolls also when the text is completely outside the canvas so it scrolls in the white.

Now i do think the textarea should follow the position until is on the html canvas element and make the page scroll, but scrolling where the canvas is not available is useless.

I can see how the scrolling part is more an app behaviour that is subject to opinions rather than a real bug - non bug.

This should be configurable in some way, but i m not certain i want to add it as a configurable options, maybe if there is a way to pass in some kind of limits for the textbox placement.

Reply:
But a click on the canvas should never trigger a page scroll ( on desktop, with no virtual keyboard )

Reply:
i also see the bad behaviour of the fiddle, but i would see it reproduced in a normal page to discuss it better. Maybe i can add a simple test case in the sandboxes of the repo

Reply:
> I understand the move page up issue. What i do not understand is if the textarea is below the fold and you don't move the page up, i tried your fiddle in a number of ways. There are 2 issues here, one is that the page scrolls and one is that it scrolls also when the text is completely outside the canvas so it scrolls in the white.
> 
> Now i do think the textarea should follow the position until is on the html canvas element and make the page scroll, but scrolling where the canvas is not available is useless.
> 
> I can see how the scrolling part is more an app behaviour that is subject to opinions rather than a real bug - non bug.
> 
> This should be configurable in some way, but i m not certain i want to add it as a configurable options, maybe if there is a way to pass in some kind of limits for the textbox placement.

I'm not sure if I understand correctly. Do you mean you want me to scroll the canvas up when the textarea is about to go below the fold?

Reply:
> i also see the bad behaviour of the fiddle, but i would see it reproduced in a normal page to discuss it better. Maybe i can add a simple test case in the sandboxes of the repo

Thank you very much. I also hope to resolve this issue better and contribute a little to this project. ^_^

Reply:
Sorry i wrote my message in a confusing way.
Let me start a couple of files to discuss this better

Reply:
No problem at all, and I'll dive deeper into the source code as well. I'm looking forward to discussing this further.